### ISSUE TITLE
Names of test folders

### ISSUE THREAD

**Issue by tisonkun** [Raku/roast#357]
Now we have folders named "SXX-<subtest-name>", which inherit from the specs. However, as time goes on, people will forget what "XX" stands for. I have an idea that we can gradually rename these folders, firstly change "S05-interpolation" into "S05-regex-interpolation", and finally "Regex-interpolation" at a right time, such like, most of folders have a more meaningful name.

Besides the renaming procedure maybe reorganize folders and files for higher readability.

What do you think of it?

**Comment by AlexDaniel** [comment]
Yes. Wonderful proposal.

Of course we will have a bunch of old-timers saying that the current situation is OK (because any change is bad and they're somewhat familiar with the numbers), but I'm hoping that the “S05-regex-interpolation” middle ground should be good enough for everyone.

**Comment by zoffixznet** [comment]
> we can gradually rename these folder [...] What do you think of it?

First, lets do this *after* 6.d release. There's a huge amount of commits to review as it is.

Second, I don't see the point in doing anything gradually. I got [3rd most commits]https://github.com/perl6/roast/graphs/contributors) to this repo and never cared for the numbers nor even the names of the folders. Just improve the structure entirely, without doing anything gradually.

And third, the real problem aren't the numbers but the multi-dimentional nature of roast packed into a flat structure. We have files for tests for a routine, tested with multiple types and inputs; we have files for tests for a type, tested with multiple routines and inputs; we have files for tests for a specific input, tested with multiple types and routines; we have files testing multiple routines, with multiple types and inputs (thematically grouped).

So I think renaming `S05-interpolation` to `Regex-interpolation` will do little to improve the structure of roast. We need a more coherent system for packing that multi-dimentional structure down and following it more strictly (an automated tool akin to [CoreHackers::Sourcery](https://github.com/zoffixznet/perl6-CoreHackers-Sourcery) that'd tell you location of tests instead of source code would be super helpful).

**Comment by W4anD0eR96** [comment]
> And third, the real problem aren't the numbers but the multi-dimentional nature of roast packed into a flat structure. We have files for tests for a routine, tested with multiple types and inputs; we have files for tests for a type, tested with multiple routines and inputs; we have files for tests for a specific input, tested with multiple types and routines; we have files testing multiple routines, with multiple types and inputs (thematically grouped).

I agree. That is why I mention "the renaming procedure maybe reorganize folders and files". Sorry for not showing main purpose directly.

**Comment by moritz** [comment]
Please note that rakudo keeps track of test files, and applies that list to older branches (like `6.c-errata`)., more to come once v6.d is out. Renaming files must be done in all branches, or it will cause quite some pain.

Finally, we need some guiding principles for where to put tests, otherwise we risk not actually improving things, just making it more intuitive for the one who does the reorganization. If we find that some implementation has a book calling `split` on complex numbers inside a loop, and we test for that, do we put the test together with the loop tests, or the `split` tests, or the complex number tests?

**Comment by zoffixznet** [comment]
> Please note that rakudo keeps track of test files, and applies that list to older branches (like 6.c-errata)., more to come once v6.d is out. Renaming files must be done in all branches, or it will cause quite some pain.

That was so in the past, but is no longer. Each branch is [driven by a separate `spectest.data` file](https://github.com/rakudo/rakudo/blob/master/docs/roast-spectest.data-versioning.md).

**Comment by jnthn** [comment]
> Of course we will have a bunch of old-timers saying that the current situation is OK (because any change is bad and they're somewhat familiar with the numbers)

@AlexDaniel, dumping on those of us who have worked on Perl 6 for a decade or more is not constructive. I read the initial suggestion by @W4anD0eR96 yesterday and was going to post something like what I'll write below. Then I saw your comment and thought, screw it, why spend a chunk of my weekend on this crap. :/

So anyway, my thoughts.

The `S\d\d` structure came from the written synopses, which in turn derived their numbers from the book Programming Perl (5), because it's - as will be discovered when trying to make things better - really hard to come up with a good organization for things, and even harder when you're still trying to figure out what's being built.

For a lot of years - in fact, almost all the way up to the 6.c release - references to the written design documents were common among those working on and early-adopting Perl 6. That's in part because they contained things that weren't covered in roast yet. While there was an aspiration for many years that the language would be specified by its test suite, not by some written documents, that was a long journey, involving creating the tens of thousands of tests we have today. When both the synopses *and* the test suite were frequently being invoked in design discussions, it was really, really useful to have a structure that we could cross-reference. Further, given the lack of documentation, many people were highly familiar with the synopses because they often had to stand in for user documentation.

Today, the situation is, happily, quite different. The journey from written to executable specification is essentially complete. I don't mean that in the sense that we don't need to write more tests - because we most certainly do - but that discussions about how things are specified today pretty much always involve looking directly at the test suite. The synopses have ceased to play much of a role there. Ditto for documentation: it's increasingly rare that we need to point anyone at the synopses, because the docs are getting to be complete enough we rarely have to.

Given all that, I don't think we have any reason to retain a structure that matches up with a set of documents that are, by now, mostly of historical interest. I agree with @zoffixznet that if we're going to do this, then we might as well use it as an opportunity for some more general restructuring. Having both an `S32-io` and an `S16-io` is one of the most obvious places that leads to pretty arbitrary decisions on where to put things. And does it really make sense for multiple dispatch tests for multi subs and multi methods to be so far apart from each other when they're the using the same mechanism underneath? It's only that way because sub stuff ended up in S06 and method stuff ended up in S12, whereas in reality we could probably understand coverage of multiple dispatch better if we had testing of given multi-dispatch features (arity, type, `where`, `is rw`, unpacks), covering the method and sub cases, together.

Also, I think what @moritz says about needing to have some guiding principles is important. We'll have to be careful not to let the perfect be the enemy of the good - because this is an area where stuff could be bikeshed all day - but we should be able to at least agree on the properties of the organization we're aiming for.

One closing thought. The tests some time back included actively maintained "smart links". These were used to render a version of the synopses with links to the tests that covered each section. These are no longer very valuable because nearly nobody reads the synopses any more. But people *do* read the language documentation. As I understand it, there's also ongoing discussions about the structure of the language documentation. It might be interesting to consider if we can structure the tests in such a way that we can cross-reference between the documentation and the tests that specify the semantics of the thing being documented.

**Comment by AlexDaniel** [comment]
TL;DR for contributors stumbling upon the wall of texts:

- [ ] Plan to change the whole structure instead of just renaming something
- [ ] An outline of principles that can be used to do the restructuring are welcome
- [ ] Would be great to link docs with tests
- [ ] Wait for v6.d release before doing the actual work

↑ feel free to edit this comment as we go.

**Comment by W4anD0eR96** [comment]
Putting forward a list of TOP LEVEL CONCEPTS might be helpful.

Now we have two tests named `io`, and `pod` things scattered here and there. The problem is we have top level concepts "Bits and Pieces" and "Setting library", which are too vague.

For these special two, we can set top level concepts "IO" and "Documentation", then gather tests.

However, things go complex when concepts are intersecting. Think of we have "Iterable" and "Operator", then where should tests of hyper operator on Array be? In this very case I think we can put it in "Operator" and mention where it is in "Iterable"(There should be one, and it is also in other place, only set a link for not repeat.) But after all, this is a typical case what we will meet.

Back to the topic, we can firstly put forward a list of TOP LEVEL CONCEPTS, like:

- IO
- Documentation
- Iterable
- Operator
- Regex
- Concurrency
...

and then deal with the intersecting problems.

**Comment by finanalyst** [comment]
A set of documents, such as the S dd series, is useful because the 
documents themselves and the index for the documents provide a way to 
navigate the complexity.

A set of principles / headings is less useful - I think - because such 
headers are open to a great deal of interpretation.

The S dd documents developed from the Synopses, which evolved out of the 
original Apocalypses and Exegeses.

The evolution was rational, even if now it seems that the final result 
may seem jumbled.

It seems to me there is a place for both concise specification documents 
(something like the Sdd, but perhaps revised), and much more expressive 
documentation for users.

A set of specification documents would be the more natural place to link 
to the tests.

However the documents/specifications/tests are ordered, if they are to 
be re-ordered elegantly, it will take a huge amount of work to get these 
documents in order. Someone is going to object. It would be better for 
there to be a consensus first about how to re-organise the 
documentation/tests, and then for the task to be undertaken in a 
systematic way as a community effort.

I say this because it seems likely some developer will take on the task 
and do it all. This will lead to friction. Just like the throw-away 
comments about 'old-timers'.

**Comment by zoffixznet** [comment]
2¢ to add: fairly often I find the need to write bug coverage tests and often the expressed bugs don't quite align with the test structure.

For example: currently writing cover for a bug with Junctions in all IO output routines. It can be covered by a single extra test, but the original tests for IO routines are split among files (e.g. S16-io/print.t, S16-io/put.t, etc.). It'd be too much work/noise to duplicate this test among all those files, so I'm gonna stick it into imperfect S16-io/print.t

Whatever the new structure is, it should keep in mind that a lot of tests are written as bug coverage tests.

**Comment by TisonKun** [comment]
Close as no longer interested in.

**Comment by AlexDaniel** [comment]
I'm still interested in this.

**Comment by tbrowder** [comment]
Note much pod is still NYI, and I rely on S26, so I hope we don't rush to hide it from those who still use it.

**Comment by tisonkun** [comment]
As https://github.com/rakudo/rakudo/issues/1222#issuecomment-346952020, close this one. Please file another to continue the discussion if any.
